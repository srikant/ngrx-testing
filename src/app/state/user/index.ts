import { createFeatureSelector, createSelector } from '@ngrx/store';
import { State } from '@state/user/user.reducer';
import * as fromUser from './user.reducer';

export const selectUserState = createFeatureSelector<State>('user');

export const selectUserIds = createSelector(
  selectUserState,
  fromUser.selectUserIds
);
export const selectUserEntities = createSelector(
  selectUserState,
  fromUser.selectUserEntities
);
export const selectAllUsers = createSelector(
  selectUserState,
  fromUser.selectAllUsers
);
export const selectUserTotal = createSelector(
  selectUserState,
  fromUser.selectUserTotal
);
export const selectSelectedUserId = createSelector(
  selectUserState,
  fromUser.getSelectedUserId
);
export const selectSelectedUser = createSelector(
  selectUserEntities,
  selectSelectedUserId,
  (entities, selectedUserId) => entities && entities[selectedUserId]
);
