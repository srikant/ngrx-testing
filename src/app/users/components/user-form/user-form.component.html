<form [formGroup]="form">
  <input type="text" placeholder="First Name" formControlName="firstName">
  <input type="text" placeholder="Last Name" formControlName="lastName">
  <button (click)="onSave()">Save</button>
</form>