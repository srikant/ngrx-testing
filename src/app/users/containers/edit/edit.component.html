<app-user-form [user]="user$ | async" (userChange)="onUserChange($event)"></app-user-form>