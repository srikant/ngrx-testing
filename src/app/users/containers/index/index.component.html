<button routerLink="add">Add User</button>
<app-user-list [users]="users | async" (selectUser)="onSelectUser($event)"></app-user-list>