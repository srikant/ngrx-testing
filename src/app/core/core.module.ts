import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import {
  ModuleWithProviders,
  NgModule,
  Optional,
  SkipSelf
} from '@angular/core';
import { HttpClientInMemoryWebApiModule } from 'angular-in-memory-web-api';
import { RootDbService } from './db/root.db.service';
import { UserService } from './services/user.service';

@NgModule({
  imports: [
    CommonModule,
    HttpClientModule,
    HttpClientInMemoryWebApiModule.forRoot(RootDbService, {
      delay: 500
    })
  ],
  declarations: [],
  providers: [UserService]
})
export class CoreModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: CoreModule
    };
  }

  constructor(
    @Optional()
    @SkipSelf()
    parentModule?: CoreModule
  ) {
    if (parentModule) {
      throw new Error(
        'CoreModule is already loaded. Import it in the AppModule only'
      );
    }
  }
}
