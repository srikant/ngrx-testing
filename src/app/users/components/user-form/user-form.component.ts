import {
  Component,
  EventEmitter,
  Input,
  OnChanges,
  Output,
  SimpleChanges
} from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { User } from '@state/user/user.model';

@Component({
  selector: 'app-user-form',
  templateUrl: './user-form.component.html',
  styleUrls: ['./user-form.component.scss']
})
export class UserFormComponent implements OnChanges {
  form: FormGroup;

  @Input() user: User;
  @Output() userChange = new EventEmitter<User>();

  constructor(private formBuilder: FormBuilder) {
    this.buildForm();
  }

  ngOnChanges(simpleChanges: SimpleChanges) {
    if (simpleChanges.user && simpleChanges.user.currentValue) {
      this.form.patchValue(simpleChanges.user.currentValue);
    }
  }

  onSave() {
    // verify form is valid
    if (this.form.invalid) {
      return;
    }

    this.userChange.emit({
      ...this.user,
      ...this.form.value
    });
  }

  private buildForm() {
    this.form = this.formBuilder.group({
      firstName: ['', Validators.required],
      lastName: ['', Validators.required]
    });
  }
}
