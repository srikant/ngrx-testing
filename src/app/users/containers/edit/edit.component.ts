import { Component, OnInit } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { Store, select } from '@ngrx/store';
import * as fromRoot from '@state/index';
import { selectSelectedUser } from '@state/user';
import { LoadUser, SelectUser, UpdateUser } from '@state/user/user.actions';
import { User } from '@state/user/user.model';
import { Observable } from 'rxjs';
import { filter, map, switchMap, tap } from 'rxjs/operators';

@Component({
  templateUrl: './edit.component.html',
  styleUrls: ['./edit.component.scss']
})
export class EditComponent implements OnInit {
  user$: Observable<User>;

  constructor(
    private activatedRoute: ActivatedRoute,
    private store: Store<fromRoot.State>
  ) {}

  ngOnInit() {
    const PARAM_ID = 'id';
    this.user$ = this.activatedRoute.paramMap.pipe(
      filter(paramMap => paramMap.has(PARAM_ID)),
      map(paramMap => paramMap.get(PARAM_ID)),
      tap(id => {
        this.store.dispatch(new SelectUser({ id: +id }));
        this.store.dispatch(new LoadUser({ id: +id }));
      }),
      switchMap(id => this.store.pipe(select(selectSelectedUser)))
    );
  }

  onUserChange(user: User) {
    this.store.dispatch(new UpdateUser({ user: user }));
  }
}
