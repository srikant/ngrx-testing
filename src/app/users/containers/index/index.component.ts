import { Component, OnInit } from '@angular/core';
import { Router } from '@angular/router';
import { Store, select } from '@ngrx/store';
import { State } from '@state/index';
import { selectAllUsers } from '@state/user';
import { LoadUsers } from '@state/user/user.actions';
import { User } from '@state/user/user.model';
import { Observable } from 'rxjs';

@Component({
  templateUrl: './index.component.html',
  styleUrls: ['./index.component.scss']
})
export class IndexComponent implements OnInit {
  users: Observable<Array<User>>;

  constructor(private router: Router, private store: Store<State>) {}

  ngOnInit() {
    this.store.dispatch(new LoadUsers());
    this.users = this.store.pipe(select(selectAllUsers));
  }

  onSelectUser(user: User) {
    this.router.navigate(['/users', user.id]);
  }
}
