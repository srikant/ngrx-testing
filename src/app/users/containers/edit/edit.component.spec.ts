import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { ActivatedRoute, convertToParamMap } from '@angular/router';
import { RouterTestingModule } from '@angular/router/testing';
import { Store } from '@ngrx/store';
import * as fromRoot from '@state/index';
import { LoadUser, SelectUser, UpdateUser } from '@state/user/user.actions';
import { User, generateUser } from '@state/user/user.model';
import { hot } from 'jasmine-marbles';
import { BehaviorSubject } from 'rxjs';
import { UserFormComponent } from './../../components/user-form/user-form.component';
import { EditComponent } from './edit.component';

describe('EditComponent', () => {
  let component: EditComponent;
  let fixture: ComponentFixture<EditComponent>;
  let store: Store<fromRoot.State>;
  let user: User;

  beforeEach(() => {
    user = generateUser();
  });

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [EditComponent, UserFormComponent],
      imports: [FormsModule, RouterTestingModule, ReactiveFormsModule],
      providers: [
        {
          provide: ActivatedRoute,
          useValue: {
            paramMap: new BehaviorSubject(
              convertToParamMap({
                id: user.id
              })
            )
          }
        },
        {
          provide: Store,
          useValue: {
            dispatch: jest.fn(),
            pipe: jest.fn(() => hot('-a', { a: user }))
          }
        }
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(EditComponent);
    component = fixture.componentInstance;
    store = TestBed.get(Store);
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  describe('ngOnInit', () => {
    it('should dispatch SelectUser action for specified id parameter', () => {
      const action = new SelectUser({ id: user.id });
      const spy = jest.spyOn(store, 'dispatch');

      fixture.detectChanges();

      expect(spy).toHaveBeenCalledWith(action);
    });

    it('should dispatch LoadUser action for specified id parameter', () => {
      const action = new LoadUser({ id: user.id });
      const spy = jest.spyOn(store, 'dispatch');

      fixture.detectChanges();

      expect(spy).toHaveBeenCalledWith(action);
    });
  });

  describe('onUserChange', () => {
    it('should dispatch the UpdateUser action when onUserChange is invoked', () => {
      const updatedUser = generateUser();
      const action = new UpdateUser({ user: updatedUser });
      const spy = jest.spyOn(store, 'dispatch');

      fixture.detectChanges();

      component.onUserChange(updatedUser);
      expect(spy).toHaveBeenCalledWith(action);
    });
  });
});
