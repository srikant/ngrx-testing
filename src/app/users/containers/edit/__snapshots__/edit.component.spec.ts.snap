// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditComponent should match snapshot 1`] = `
<ng-component
  activatedRoute={[Function Object]}
  store={[Function Object]}
  user$={[Function AnonymousSubject]}
>
  <app-user-form>
    <form
      class="ng-untouched ng-pristine ng-invalid"
      ng-reflect-form="[object Object]"
      novalidate=""
    >
      <input
        class="ng-untouched ng-pristine ng-invalid"
        formcontrolname="firstName"
        ng-reflect-name="firstName"
        placeholder="First Name"
        type="text"
      />
      <input
        class="ng-untouched ng-pristine ng-invalid"
        formcontrolname="lastName"
        ng-reflect-name="lastName"
        placeholder="Last Name"
        type="text"
      />
      <button>
        Save
      </button>
    </form>
  </app-user-form>
</ng-component>
`;
